### PR TITLE
(feat): Prompt engineering to remind o1 to generate a patch

### PR DESCRIPTION
**Pull request by AlexCuadron**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

O1 is very forgetful, it usually stops coding as soon as it replicates the issue.
Modified the environment reminder so that o1 remembers to output only _one action_ and to _only_ press finish whenever it has generated a patch.

---
**Link of any specific issues this addresses**

### CODE REVIEW

**Comment by enyst** [comment]
It would be good if @xingyaoww can take a look at it.

Currently, I believe we don't have function calling with o1. That might change, and then function calling allows multiple actions. Maybe that's okay, we just may need to adjust the other part in the prompt there.

**Comment by AlexCuadron** [comment]
> It would be good if @xingyaoww can take a look at it.
> 
> Currently, I believe we don't have function calling with o1. That might change, and then function calling allows multiple actions. Maybe that's okay, we just may need to adjust the other part (the finish part) in the prompt for function calls.

FC is not available from the [API side](https://openai.com/index/introducing-openai-o1-preview/), the biggest issue here is forgetting to generate a patch. Most of the `git diffs` generated by o1-prev or o1-mini are just `reproduce_issue.py`

File: openhands/agenthub/codeact_agent/codeact_agent.py [hunk 1]
```diff
@@ -509,7 +509,16 @@ def _get_messages(self, state: State) -> list[Message]:
...
             if latest_user_message:
-                reminder_text = f'\n\nENVIRONMENT REMINDER: You have {state.max_iterations - state.iteration} turns left to complete the task. When finished reply with <finish></finish>.'
+                reminder_text = (
+                    '\n\nENVIRONMENT REMINDER:\n'
+                    f'- You have {state.max_iterations - state.iteration} turns left to complete the task\n'
+                    '- When finished reply with <finish></finish>\n'
+                    '\n<IMPORTANT>\n'
+                    '- You MUST generate only one action per turn!\n'
+                    '- A patch is a set of changes to the source code of the codebase that you are given\n'
+                    '- You MUST generate a patch that attempts to fix the issue described in the <pr_description>\n'
```
Comment (xingyaoww) [inline openhands/agenthub/codeact_agent/codeact_agent.py:519]: Can we make this only available for SWE-Bench evaluation? This seems to be less useful for other general tasks?

Other than that i think I'm overall good with this change!

**Comment by AlexCuadron** [comment]
Fixed! PTAL @xingyaoww

**Review by xingyaoww — COMMENTED** [review]
Hmm giving it a second thought, i'm actually wondering if we can do this inside `run_infer.py` -- Can we just add this as a suffix in the initial instruction?

The reason is that the environment reminder will likely go away as i got this https://github.com/All-Hands-AI/OpenHands/pull/4711 working -- then everything will be organized as "function calling" code

**Comment by AlexCuadron** [comment]
Fixed! PTAL, @xingyaoww

File: evaluation/swe_bench/run_infer.py [hunk 2]
```diff
@@ -91,6 +91,14 @@ def get_instruction(instance: pd.Series, metadata: EvalMetadata):
             "Your thinking should be thorough and so it's fine if it's very long.\n"
         )
 
+    instruction += """
+        <IMPORTANT>
```
Comment (xingyaoww) [inline evaluation/swe_bench/run_infer.py:95]: Should we remove the indentation before each line?

File: openhands/agenthub/codeact_agent/codeact_agent.py [hunk 3]
```diff
@@ -508,8 +508,12 @@ def _get_messages(self, state: State) -> list[Message]:
                 None,
             )
             # do not add this for function calling
-            if latest_user_message:
-                reminder_text = f'\n\nENVIRONMENT REMINDER: You have {state.max_iterations - state.iteration} turns left to complete the task. When finished reply with <finish></finish>.'
-                latest_user_message.content.append(TextContent(text=reminder_text))
+            if not latest_user_message:
+                return messages
```
Comment (xingyaoww) [inline openhands/agenthub/codeact_agent/codeact_agent.py:512]: Shall we revert this part to how it looks like before?

**Review by xingyaoww — COMMENTED** [review]
A few nit - but overall LGTM!

**Comment by AlexCuadron** [comment]
Fixed! PTAL again :D @xingyaoww

**Review by xingyaoww — APPROVED** [review]
LGTM! Thanks!

**Comment by xingyaoww** [comment]
@AlexCuadron After giving it more thought - i think maybe we should revert or revise this -- basically the agent SHOULD NOT be aware of the patch directly, instead, they should simply interact with the environment to edit files & let us know when it finished - and we will grab a patch from `git diff`.

```
<IMPORTANT>
- You MUST generate only one action per turn!
- A patch is a set of changes to the source code of the codebase that you are given
- You MUST generate a patch that attempts to fix the issue described in the <pr_description>
</IMPORTANT>
```

**Comment by xingyaoww** [comment]
Can you share more info on how O1's behavior will be when this is not provided? I'd assume we maybe able to tweak other parts of the instruction to get it working instead of explicitly mention "patch"

**Comment by enyst** [comment]
Maybe we can rephrase it as "MUST solve the task".

**Comment by xingyaoww** [comment]
@enyst I think it ultimately comes down to the eventual effectiveness -- "MUST solve the task" might not work very well of o1 :( - I guess we can wait for @AlexCuadron's testing in terms of what works and what doesn't, and we can figure out from there

**Comment by AlexCuadron** [comment]
O1 is not very good at instruction following, it needs a reminder at the end. Otherwise, it thinks that after finishing reproducing the issue it's done and it generates a `<finish></finish>`

**Pull request merged**